### PR TITLE
complete refactor

### PR DESCRIPTION
**Pull request by jonschlinkert**
_(~~NOTE: bash-test-related code is causing tests to fail, I'm working on getting that removed~~)_

Complete overhaul, with 36,000+ new unit tests validated against actual output generated by Bash and minimatch. More specifically, 35,000+ of the tests:

- micromatch results are directly compared to bash results
- in rare cases, when micromatch and bash disagree, micromatch's results are compared to minimatch's results
- micromatch is much more accurate than minimatch, so there were cases where I had to make assumptions. I'll try to document these.

This refactor introduces a parser and compiler that are supersets of more granular parsers and compilers from other sub-modules. Each of these sub-modules has a singular responsibility and focuses on a certain type of matching that aligns with a specific part of the Bash "expansion" API. 

These sub-modules work like plugins to seamlessly create the micromatch parser/compiler, so that strings are parsed in one pass, an [AST is created](https://gist.github.com/jonschlinkert/099c8914f56529f75bc757cc9e5e8e2a), then a new string is generated by the compiler.

Here are those sub-modules with links to related prs on those modules if you want to see how they contribute to this code:
- [nanomatch](https://github.com/jonschlinkert/nanomatch) (new library) - glob expansion (`*`, `**`, `?` and `[...]`))
- [braces](https://github.com/jonschlinkert/braces/pull/10) - brace expansion (`{1..10}`, `{a,b,c}`, etc)
- [extglob](https://github.com/jonschlinkert/extglob/pull/5) - extended globs (`!(a|b)`, `@(!(foo|bar))`, etc)
- [expand-brackets](https://github.com/jonschlinkert/expand-brackets/pull/5) - POSIX character classes `[[:alpha:][:digit:]]`

**Additional features**
- parser is exposed, so that implementors can customize or override specific micromatch parsers if necessary
- compiler is exposed, so that implementors can customize or override specific micromatch compilers if necessary
- more accurate matching (passes 100% of Bash 4.3 of the brace expansion and extglob unit tests, as well as all Bash glob tests that are relevant to node.js usage, all minimatch tests, all brace-expansion tests, and also passes a couple of tests that bash fails)
- even safer - micromatch has always generated optimized patterns so it's not subject to DoS exploits like minimatch (completely different than the regex DoS issue, minimatch and multimatch are still openly exposed to being used for DoS attacks), but more [safeguards were built into this refactor](https://github.com/jonschlinkert/braces#braces-is-safe)
- source map support (optionally created when using parse or compile - I have no idea what the use case is yet, but they come for free) (note that source maps are not generated for brace expansion at present, since the braces compiler uses a different strategy. I'll update if/when this changes)

**Example source map**

Just for fun, again no idea what this will be used for lol

``` js
var mm = require('micromatch');
var pattern = '*(*(of*(a)x)z)';

var ast = mm.parse(pattern, {sourcemap: true});
var res = mm.compile(ast);
console.log(res);
// { map:
//    { version: 3,
//      sources: [ 'string' ],
//      names: [],
//      mappings: 'AAAA,CAAE,CAAE,EAAE,CAAE,CAAC,EAAC,CAAC,EAAC,CAAC,EAAC',
//      sourcesContent: [ '*(*(of*(a)x)z)' ] }},
```

Still working on readme/docs. Any and all feedback, or code reviews would be appreciated.

cc @doowb @phated @hemanth @es128 @eush77 @paulmillr @tunnckoCore @TrySound

### CODE REVIEW

**Comment by tunnckoCore** [comment]
Everything is okey to me :P It's awesome as usual. I just don't like that new style of `example` and `params` order in readme API. I mean, the example of a method is first then is the "params" of a method. I feel it uncomfortable to be and looks not so natural. Also it seems it is intentional, because I seen the same thing in some other repos from last days. I believe this effect is done by not adding empty line after the closing code fence (aka before the first `@param`).

It's not so hard thing, but just share my feelings and that i don't like this direction of readmes ;d

**Comment by jonschlinkert** [comment]
> I just don't like that new style of example and params order in readme API.

I don't think I like it either. It seemed nice at first. But let's not discuss this here, we can continue this discussion on verb-generate-readme if you'd like to give more feedback. thanks

File: .travis.yml [hunk 1]
```diff
@@ -1,12 +1,22 @@
 sudo: false
+os:
+  - linux
+  - osx
 language: node_js
 node_js:
+  - node
   - '6'
   - '5'
-  - '4'
```
Comment (jonschlinkert) [inline .travis.yml]: yes this makes sense. I have updated the templates that generate/update this file. thanks for the reminder

Comment (es128) [inline .travis.yml]: 0.10 was the mainstream version for a long time, and many enterprise users are still stuck on it. We want to keep supporting it for a while.

**Comment by jonschlinkert** [comment]
we will always find edge cases, but no one can say we didn't have enough unit tests lol

screen shot 2016-12-05 at 6 13 56 pm

**Comment by tunnckoCore** [comment]
@jonschlinkert hxahhahahahah, absolutely! You're unreal, srsly :100: 

Can't wait it more.

File: .gitignore [hunk 2]
```diff
@@ -18,3 +19,4 @@ vendor
 temp
 tmp
 TODO.md
+yarn.lock
```
Comment (tunnckoCore) [inline .gitignore]: it should not be ignored. it should exists in the repositories. https://github.com/yarnpkg/yarn/issues/963

**Comment by wtgtybhertgeghgtwtg** [comment]
There's a lot of deduplication issues with the dependencies; do you want me to open an issue or file some PR's or something?

**Comment by jonschlinkert** [comment]
> There's a lot of deduplication issues with the dependencies;

Only if you've checked deps for API changes or potential incompatibilities and you know that there won't be any regressions from upgrading. E.g. not all changes are bugfixes.  Although it would probably make the most sense after this branch was published, since updating deps is something I already do as an automatic part of my pre-publish process.

**Comment by wtgtybhertgeghgtwtg** [comment]
The problem is transitive dependencies (i.e. `nanomatch` and `braces` depend on different versions of `snapdragon`, and `snapdragon` has two versions of `isobject` in its dependency tree).  So `braces` would have to bump its version of `snapdragon` to match the version `nanomatch` uses, and a handful of things in the dependency tree of `snapdragon` would have to bump their version of `isobject`.
If you've got a script that sorts that out over however many packages, that sounds pretty cool.

**Comment by wtgtybhertgeghgtwtg** [comment]
https://github.com/jonschlinkert/snapdragon/issues/6 has been filed for deduping `snapdragon`.  If it is acceptable, I will look into resolving that.

**Comment by jonschlinkert** [comment]
@wtgtybhertgeghgtwtg since you created an issue on snapdragon for us to discuss,  I'm going to delete the previous two comments to keep this issue focused on the code in the pull request. we can continue the discussion on snapdragon.

**Comment by phated** [comment]
@jonschlinkert saw all the recent commits come in on this.  Looks awesome, pretty close to release?

**Comment by jonschlinkert** [comment]
> Looks awesome, pretty close to release?

thanks! Yeah, I think it's ready I just need to finish docs. Shouldn't take much longer, just formatting the docs to be more [like this](https://github.com/jonschlinkert/nanomatch#documentation).

**Comment by phated** [comment]
@jonschlinkert I like those docs but it'd be nice if the top option in the API section were expanded by default (you almost always want to know that one).

**Comment by jonschlinkert** [comment]
> I like those docs but it'd be nice if the top option in the API section were expanded by default (you almost always want to know that one).

the main export you mean? yeah, that makes a lot of sense. this is a new format I'm experimenting with, there is always room for improvement.

**Comment by jonschlinkert** [comment]
@phated before I update the readme to add those sections, do you like that format? I'm interested in feedback from others if anyone has an opinion. The main reason I was doing it is that I've seen a couple of comments from devs feeling "overwhelmed" by long readmes. I personally like long docs, as there is no rule forcing me to read them all, and they're there when I need them.

**Comment by phated** [comment]
@jonschlinkert I like what you did in nanomatch.  I think the best balance would be to collapse any docs that are not-often-used or you deem to not be very important (such as most options, etc).  Anything that might trip people up or are critical to usage should be expanded.

**Comment by jonschlinkert** [comment]
> I think the best balance would be to collapse any docs that are not-often-used or you deem to not be very important

k, maybe a good approach then would be to just point those things out with links and/or additional examples. Might make it harder to read if some sections are collapsed and other expanded. Let's just view it as a living document, I'll give it a go, then we can change whatever is needed. fwiw, removing the sections is easy, it's just a setting in verb.

**Comment by jonschlinkert** [comment]
@phated also I'm thinking of moving all of the *match libs and any related libs to an org (I think `github.com/micromatch`). I'll add you and anyone else that is already a collab on the orgs. 

Any objections from anyone on that? @es128, @doowb, anyone else?

**Comment by phated** [comment]
Woot!  Orgs rock!

**Comment by es128** [comment]
👍 Sounds good
